#8525 Add SQL Branch Operator

samuelkhtu opened the pull request:
SQL Branch Operator allow user to execute a SQL query in any supported backend to decide which
branch of the DAG to follow. 

The SQL branch operator expects SQL query to return any of the following:
- Boolean: True/False 
- Integer: 0/1 
- String: true/y/yes/1/on/false/n/no/0/off

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

Commit:
Add SQL Branch Operator

SQL Branch Operator allow user to execute a SQL query in any supported backend to decide which
branch to follow. The SQL branch operator expect query to return True/False (Boolean) or
0/1 (Integer) or true/y/yes/1/on/false/n/no/0/off (String).

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,165 @@
+            self.parameters,
+            self._hook,
+        )
+        records = self._hook.get_records(self.sql, self.parameters)
```
why `get_records`? are you expecting to more than 1 row?
Shouldn't it use `get_first`?
```suggestion
        records = self._hook.get_first(self.sql, self.parameters)
```

samuelkhtu replied:
Yes, Thank you. This make perfect sense.

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,165 @@
+    def __init__(
+        self,
+        sql: str,
+        follow_task_ids_if_true: str,
```
it can be list of str if you have more than one task

samuelkhtu replied:
Thanks, I will update the type hints.

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,165 @@
+        self,
+        sql: str,
+        follow_task_ids_if_true: str,
+        follow_task_ids_if_false: str,
```
same

samuelkhtu replied:
Absolutely

eladkal reviewed (changes requested):
I think this is very useful!

Tested several scenarios with Presto. Works as expected.
I think it's worth also changing the background color. It's very hard to read black text with bold red background
![Screen Shot 2020-05-21 at 9 59 50](https://user-images.githubusercontent.com/45845474/82535540-a9b2a400-9b4f-11ea-8b9e-14bf0711f004.png)
![Screen Shot 2020-05-21 at 10 44 40](https://user-images.githubusercontent.com/45845474/82535813-2180ce80-9b50-11ea-9ef1-adc36d79156a.png)

Commit:
Resolve issues found during PR review

1. Modify foreground color to be more visible.
2. Replace hook.get_records with hook.get_first as SQL branch operator expects single row return.
3. Modify follow_task_ids_if_true/false parameters with List[str] type hint instead of str.

Commit:
Add SQL branch operator to operators and hooks reference document

Commit:
Fix document blank line formatting issue

Commit:
Fix SQL return record data type error

1. Add tuple and list return data type to SQL branch operator.
2. Add unit test for single return value

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,175 @@
+            )
+
+        if isinstance(record, list):
+            query_result = record[0][0]
```
I get an error related to this line
```
  File "/home/airflow/dags/testbranch.py", line 126, in execute
    query_result = record[0][0]
TypeError: 'bool' object is not subscriptable
```

Did it work for you when you tested against other DBs?

samuelkhtu replied:
@eladkal  Thank you for review again. This is unexpected. I found an issue with the unit test and fixed this issue as well. The PR should be ready to review again.

Commit:
Fix SQL branch operator query result handle

1. Query get_first should return row['column'] instead.
2. Fixed unit test to return list of result instead of list of list of result.

eladkal commented on tests/operators/test_sql_branch_operator.py:
```diff
@@ -0,0 +1,482 @@
+
+# import os
+# import unittest
+# from unittest import mock
```
no need for that

samuelkhtu replied:
Roger that! Removed

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,178 @@
+    :param parameters: (optional) the parameters to render the SQL query with.
+    :type parameters: mapping or iterable
+
+
```
Not sure you need to have blank lines between each params. check other operators, for example:
https://github.com/apache/airflow/blob/1d36b0303b8632fce6de78ca4e782ae26ee06fea/airflow/providers/mysql/operators/mysql.py#L27

samuelkhtu replied:
Removed blank lines

eladkal commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,178 @@
+        if self.follow_task_ids_if_false is None:
+            raise AirflowException(
+                "Expected task id or task ids assigned to follow_task_ids_if_false"
+            )
```
If you check for mandatory arguments why not check also for `sql` param?

samuelkhtu replied:
Sounds good. Added check for SQL parameter as well.

eladkal reviewed (commented):
overall LGTM left few minor comments

tested with Presto db
![Screen Shot 2020-05-24 at 11 50 51](https://user-images.githubusercontent.com/45845474/82749945-5e99ca80-9db5-11ea-9a94-5fdde389c1e6.png)

Commit:
Remove unused import and add additional parameter check

1. Remove unused import from unit test file.
2. Add check to verify required SQL paramter.
3. Remove blank line between param and parm type docstring.

eladkal reviewed (approved):
LGTM.

One last observation about the file name `sql_branch_operator.py`. By accepting this PR Airflow will have 2 sql related operator files in the core (`check_operator.py`, `sql_branch_operator.py`) so maybe the file name should be changed `sql_branch_operator.py` -> `sql.py` so that In the future [check_operator.py](https://github.com/apache/airflow/blob/master/airflow/operators/check_operator.py) can be deprecated by move the operators into `sql.py` Similar to the deprecation of [python_operator.py](https://github.com/apache/airflow/blob/master/airflow/operators/python_operator.py) by moving classes to [python.py](https://github.com/apache/airflow/blob/master/airflow/operators/python.py)
Lets wait to see what others think of that.

potiuk commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,174 @@
+        self.log.debug("Get connection for %s", self.conn_id)
+        conn = BaseHook.get_connection(self.conn_id)
+
+        allowed_conn_type = {
```
Should you make it CONSTANT?

samuelkhtu replied:
I moved and converted the "allowed_conn_type" variable to ALLOWED_CONN_TYPE constant. I also moved the constant outside of the class. What do you think? Thanks

```Python
ALLOWED_CONN_TYPE = {
    "google_cloud_platform",
    "jdbc",
    "mssql",
    "mysql",
    "odbc",
    "oracle",
    "postgres",
    "presto",
    "sqlite",
    "vertica",
}


class BranchSqlOperator(BaseOperator, SkipMixin):

potiuk commented on airflow/operators/sql_branch_operator.py:
```diff
@@ -0,0 +1,174 @@
+    def execute(self, context: Dict):
+        # get supported hook
+        self._hook = self._get_hook()
+
```
Maybe it's a good idea to check if the hook is DbApiHook instance? I think it's even better check than checking the connection types.

samuelkhtu replied:
Hi @potiuk , If I added the following code to check if the hook is DbApiHook instance, my unit test's mock will break. Any suggestion? I searched the Airflow code base but I couldn't find an example to resolve this issue. Thanks!

```Python
if not isinstance(self._hook, DbApiHook):
    raise AirflowException(
        "Unexpected type returned '%s' expected DbApiHook" % type(self._hook))

# In unit test, my test will fail with
"airflow.exceptions.AirflowException: Unexpected type returned '<class 'unittest.mock.MagicMock'>' expected DbApiHook"
```

potiuk replied:
Just rebase it to latest master please. We had some problems with kubernetes tests - they should be largely solved now (I am working on final fix) but at least it should not fail now.

potiuk replied:
Here is some info on that: https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#how-to-rebase-pr

samuelkhtu replied:
Hi @potiuk , I haven't started the breeze integrated test yet. I only run the unit test locally. So, I think this is a "mock" issue.

```Python
pytest ../airflow/tests/operators/test_sql_branch_operator.py
```
In the unit test, I used mock connection (Similar to the existing unit "test test_sql_sensor")
``` Python
mock_hook.get_connection("mysql_default").conn_type = "mysql"
```
Therefore, during my local pytest run, the new "isinstance(self._hook, DbApiHook)" check failed and return error.

```Python
def execute(self, context: Dict):
        # get supported hook
        self._hook = self._get_hook()
    
        if not isinstance(self._hook, DbApiHook):
            raise AirflowException(
>               "Unexpected type returned '%s' expected DbApiHook" % type(self._hook)
            )
E           airflow.exceptions.AirflowException: Unexpected type returned '<class 'unittest.mock.MagicMock'>' expected DbApiHook
```

samuelkhtu replied:
I also checked other operators to see if any of them verify the DbApiHook return type but I couldn't find any example of that check.

potiuk replied:
OK. Fine for me. Indeed that is quite OK this way :)

potiuk replied:
Can you please rebase to latest master? That should fix the failure problem.

samuelkhtu replied:
Thank you so much for reviewing this PR. Yes, I will rebase my PR. Thanks!

potiuk reviewed (commented):
Looks really great and sounds super useful - just two nits -if you can take a look

Commit:
Change allowed_conn_type to constant

eladkal commented on tests/operators/test_sql_branch_operator.py:
```diff
@@ -0,0 +1,479 @@
+            default_args={"owner": "airflow", "start_date": DEFAULT_DATE},
+            schedule_interval=INTERVAL,
+        )
+        # self.dag = DAG(TEST_DAG_ID, default_args=args)
```
do we need this comment?

samuelkhtu replied:
Comment removed. I also removed unused TEST_DAG_ID constant from the test. Thanks.

samuelkhtu committed:
Remove unused comments and variables in unit test
